src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs

```diff
@@ -435,14 +436,124 @@ private X509Certificate2 LoadCertificate(CertificateConfig certInfo, string endp
             }
             else if (certInfo.IsFileCert)
             {
-                var env = Options.ApplicationServices.GetRequiredService<IHostEnvironment>();
-                return new X509Certificate2(Path.Combine(env.ContentRootPath, certInfo.Path), certInfo.Password);
+                var environment = Options.ApplicationServices.GetRequiredService<IHostEnvironment>();
+                var certificatePath = Path.Combine(environment.ContentRootPath, certInfo.Path);
+                if (certInfo.KeyPath != null)
+                {
+                    var certificateKeyPath = Path.Combine(environment.ContentRootPath, certInfo.KeyPath);
+                    var certificate = GetCertificate(certInfo, certificatePath, certificateKeyPath);
+
+                    if (certificate != null)
+                    {
+                        certificate = LoadCertificateKey(certificate, certificateKeyPath, certInfo.Password);
```

Comment:
It looks like LoadCertificateKey ever return null if the original cert doesn't have one of the 3 expected OIDs. We should log if this happens.

It also looks like the following call to PersistKey will nullref, so we should probably short-circuit in that case.

Reply:
You are totally right, this was an oversight on my part. I've fixed it and added logging to capture the scenarios in which a cert can be null for some reason.